src/group_theory/free_product.lean

```diff
@@ -0,0 +1,239 @@
+
+-/
+
+variables {ι : Type*} (M : Π i : ι, Type*) [Π i, monoid (M i)]
+
+/-- A relation on the free monoid on alphabet `Σ i, M i`, relating `⟨i, 1⟩` with `1` and
+`⟨i, x⟩ * ⟨i, y⟩` with `⟨i, x * y⟩`. -/
+inductive free_product.rel : free_monoid (Σ i, M i) → free_monoid (Σ i, M i) → Prop
+| of_one (i : ι) : free_product.rel (free_monoid.of ⟨i, 1⟩) 1
+| of_mul {i : ι} (x y : M i) : free_product.rel (free_monoid.of ⟨i, x⟩ * free_monoid.of ⟨i, y⟩)
+  (free_monoid.of ⟨i, x * y⟩)
+
+/-- The free product (categorical coproduct) of an indexed family of monoids. -/
+@[derive [monoid, inhabited]]
+def free_product : Type* := (con_gen (free_product.rel M)).quotient
+
+namespace free_product
+
+/-- The type of reduced words. A reduced word cannot contain a letter `1`, and no two adjacent
+letters can come from the same summand. -/
+def word : Type* := { l : list (Σ i, { m : M i // m ≠ 1 }) // (l.map sigma.fst).chain' (≠) }
+
+variable {M}
+
+/-- The inclusion of a summand into the free product. -/
+def of {i : ι} : M i →* free_product M :=
+{ to_fun   := λ x, con.mk' _ (free_monoid.of $ sigma.mk i x),
+  map_one' := (con.eq _).mpr (con_gen.rel.of _ _ (free_product.rel.of_one i)),
+  map_mul' := λ x y, eq.symm $ (con.eq _).mpr (con_gen.rel.of _ _ (free_product.rel.of_mul x y)) }
+
+/-- A map out of the free product corresponds to a family of maps out of the summands. This is the
+universal property of the free product, charaterizing it as a categorical coproduct. -/
+def lift {N} [monoid N] : (Π i, M i →* N) ≃ (free_product M →* N) :=
+{ to_fun := λ fi, con.lift _ (free_monoid.lift $ λ p : Σ i, M i, fi p.fst p.snd)
+    begin
+      refine con.con_gen_le _,
+      rintros p q (i | ⟨i, x, y⟩),
+      { simp only [free_monoid.lift_apply, con.ker_rel, con.rel_eq_coe, mul_one, list.prod_cons,
+          list.prod_nil, list.map, monoid_hom.map_one], },
+      { simp only [free_monoid.of, free_monoid.lift_apply, con.ker_rel, con.rel_eq_coe, mul_one,
+          list.append, monoid_hom.map_mul, list.prod_cons, list.prod_nil, list.map], }
+    end,
+  inv_fun := λ f i, f.comp of,
+  left_inv := λ fi, by { ext i x, simp only [of, free_monoid.lift_eval_of, function.comp_app,
+    monoid_hom.coe_mk, monoid_hom.coe_comp, con.lift_mk'], },
+  right_inv := λ f, (monoid_hom.cancel_right con.mk'_surjective).mp $ free_monoid.hom_eq $ λ ⟨i, x⟩,
+    by simp only [of, free_monoid.lift_eval_of, con.lift_comp_mk', monoid_hom.coe_mk,
+      monoid_hom.coe_comp], }
+
+lemma induction_on {C : free_product M → Prop}
+  (m : free_product M)
+  (h_one : C 1)
+  (h_of : ∀ {i} (m : M i), C (of m))
+  (h_mul : ∀ {x y}, C x → C y → C (x * y)) :
+  C m :=
+begin
+  apply con.induction_on m,
+  intro m',
+  apply m'.rec_on,
+  { rw con.coe_one, exact h_one, },
+  { rintros ⟨i, x⟩ m'' ih, rw con.coe_mul, exact h_mul (h_of x) ih },
+end
+
+noncomputable instance (G : ι → Type*) [∀ i, group (G i)] : group (free_product G) :=
+group_of_is_unit $ λ m,
```

Comment:
Is this obtainable computably with `lift (of.comp (inv_hom))`, where `inv_hom : G →* Gᵒᵖ` exists somewhere?

Reply:
Or at least, something similar to the strategy used by `clifford_algebra.reverse`

Reply:
Yes, we should be able to make `inv` computable by defining it using `lift` in this way. Good idea.

Reply:
(Another way would be to prove a group analogue of `category_theory.groupoid.of_trunc_split_mono`.)